### PR TITLE
Fix CMake, and build flags (addresses some previously inadvertently suppressed warnings) 

### LINKED ISSUE #17: CXX_FLAGS are overwritten in lib/CMakeLists.txt
This causes `-fopenmp -mavx -O3 -DNDEBUG` to overwrite a lot of other default flags, including `-Wall` which has supressed a ton of warnings that ultimately cause issues for OSX.

### PR DESCRIPTION
**Pull request by jacobagilbert**
Modernizes CMake and fixes a significant bug related to CXX_FLAGS being overwritten.

Also addresses a number of warnings.

Closes #17

### CODE REVIEW

File: lib/CMakeLists.txt [hunk 1]
```diff
@@ -44,7 +44,7 @@ if(APPLE)
     )
 endif(APPLE)
 
-set(CMAKE_CXX_FLAGS "-fopenmp -mavx -fext-numeric-literals")
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fopenmp -mavx")
```
Comment (jacobagilbert) [inline lib/CMakeLists.txt]: This is the most important fix.

**Comment by samwhiting-snl** [comment]
The CMake Flags fix is definitely needed, and probably is missing in our other repos as well. The `fext-numeric-literals` was required for our builds on centos and Ubuntu in order to use `j` in complex numbers, maybe we should combine your two Pull requests into one so we can keep this.

```
unable to find numeric literal operator ‘operator""j’
```

File: lib/CMakeLists.txt [hunk 2]
```diff
@@ -44,7 +44,7 @@ if(APPLE)
     )
 endif(APPLE)
 
-set(CMAKE_CXX_FLAGS "-fopenmp -mavx -fext-numeric-literals")
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fopenmp -mavx -fext-numeric-literals")
```
Comment (jacobagilbert) [inline lib/CMakeLists.txt:47]: This is no longer removed in this PR, so it should be mergable.

I don't see that error on clang (osx)

File: lib/CMakeLists.txt (hunk 1 above)
Comment (jacobagilbert) [reply lib/CMakeLists.txt]: OBE ☝️

**Pull request merged**